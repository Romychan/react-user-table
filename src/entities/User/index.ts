export {
  useFetchUsers,
  useFetchUser,
  useCreateUser,
  useDeleteUser,
  useUpdateUser,
} from './api/queries';
export { userHandlers } from './api/msw/userHandlers';
export type { UsersResponse, UserDTO } from './api/types';
export { MOCK_USER } from './api/__mocks__/mocks';

export { UserContextProvider, UserContext } from './model/context';
export { useUserStore } from './model/hooks';
export {
  TABLE_USERS_COLUMNS,
  DEFAULT_USERS_PARAMS,
  FORM_USERS_FIELDS,
} from './model/constants';
export type { User, UserQueryParams, UserForm } from './model/types';
export { UserFormSchema } from './model/schemas';

export { UserAvatar } from './ui/UserAvatar/UserAvatar';
