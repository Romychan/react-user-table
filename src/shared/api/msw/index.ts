export { initMockDatabase, resetMockDatabase } from './database';
