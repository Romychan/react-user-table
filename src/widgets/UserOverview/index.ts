export { UserOverview } from './ui/UserOverview';
