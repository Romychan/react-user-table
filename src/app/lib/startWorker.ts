import { userHandlers } from '~/entities/User';

/**
 * A function to start the MSW worker to intercept requests
 *
 * @returns Promise with worker
 */
export const startWorker = async () => {
  const { worker } = await import('~/shared/api/msw/browser');
  worker.use(...userHandlers);

  return worker.start({ onUnhandledRequest: 'bypass' });
};
