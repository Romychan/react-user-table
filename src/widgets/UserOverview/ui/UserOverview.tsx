import { UpdateUser } from '~/features/UpdateUser';
import { DeleteUser } from '~/features/DeleteUser';

import { useUserStore, useFetchUser } from '~/entities/User';

import { Drawer } from '~/shared/ui/Drawer';
import { Loader } from '~/shared/ui/Loader';

import styles from './UserOverview.module.scss';

/** This component is used to view and edit the user information */
export const UserOverview = () => {
  const { currentUserId, setCurrentUserId } = useUserStore();

  const { data, isLoading } = useFetchUser(currentUserId);

  return (
    <Drawer
      isVisible={!!currentUserId}
      className={styles.container}
      position="right"
      title="Overview"
      onClose={() => setCurrentUserId({ id: null })}
    >
      {(onClose) =>
        data && !isLoading ? (
          <UpdateUser
            user={data}
            actionSlot={<DeleteUser id={currentUserId} onComplete={onClose} />}
          />
        ) : (
          <Loader className={styles.loader} />
        )
      }
    </Drawer>
  );
};
