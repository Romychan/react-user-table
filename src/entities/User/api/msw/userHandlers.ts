import { HttpResponse, delay, http } from 'msw';
import { v4 as uuidv4 } from 'uuid';

import { initMockDatabase } from '~/shared/api/msw';
import { CONFIG } from '~/shared/config';

import { UserDTO } from '../types';

const database = initMockDatabase();

/** An array of `MSW` handlers for intercepting user CRUD requests */
export const userHandlers = [
  http.get(`${CONFIG.API_URL}/users`, async ({ request }) => {
    const url = new URL(request.url);

    const skip = Number(url.searchParams.get('skip'));
    const limit = Number(url.searchParams.get('limit'));
    const sort = url.searchParams.get('sort') as keyof UserDTO;
    const order = url.searchParams.get('order') as 'asc' | 'desc';
    const search = url.searchParams.get('query') as string;

    const users = database.user.findMany({
      where: {
        name: { contains: search },
      },
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      orderBy: {
        [sort]: order,
      },
      take: limit,
      skip: skip,
    });

    const total = database.user.count({
      where: {
        name: { contains: search },
      },
    });

    await delay();

    return HttpResponse.json({ users, total, skip, limit }, { status: 200 });
  }),

  http.post(`${CONFIG.API_URL}/users`, async ({ request }) => {
    const body = (await request.json()) as UserDTO;

    const newUser = {
      ...body,
      id: uuidv4(),
    };

    const user = database.user.create(newUser);

    return HttpResponse.json(user, { status: 200 });
  }),

  http.put(`${CONFIG.API_URL}/users/:id`, async ({ params, request }) => {
    const { id } = params;
    const body = (await request.json()) as UserDTO;

    const user = database.user.update({
      where: { id: { equals: id as string } },
      data: body,
    });

    return HttpResponse.json(user, { status: 200 });
  }),

  http.delete(`${CONFIG.API_URL}/users/:id`, ({ params }) => {
    const { id } = params;

    const user = database.user.delete({
      where: { id: { equals: id as string } },
    });

    return HttpResponse.json(user, { status: 200 });
  }),

  http.get(`${CONFIG.API_URL}/users/:id`, async ({ params }) => {
    const { id } = params;

    const user = database.user.findFirst({
      where: { id: { equals: id as string } },
    });

    await delay();

    return HttpResponse.json(user, { status: 200 });
  }),
];
