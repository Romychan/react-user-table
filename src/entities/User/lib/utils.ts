import { UserDTO } from '../api/types';
import { User, UserQueryParams } from '../model/types';

/**
 * A function for transforming a user object from a `UserDTO` model to a `User`
 *
 * @param user An object with user data from the server
 *
 * @returns The transformed object with user data
 */
export const mapUser = (user: UserDTO): User => {
  return {
    id: user.id,
    name: user.name,
    image: user.image,
    username: user.username,
    email: user.email,
    phone: user.phone,
    city: user.city,
    address: user.address,
  };
};

/**
 * A function for getting an API endpoint with a list of users
 *
 * @param queryParams Query parameters for the endpoint
 *
 * @returns API endpoint for getting a list of users
 */
export const getUsersAPIEndpoint = (queryParams: UserQueryParams) =>
  `/users?${new URLSearchParams({ ...queryParams }).toString()}`;
