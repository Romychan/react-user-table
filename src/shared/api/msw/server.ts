import { setupServer } from 'msw/node';

/** Set up requests interception worker in the server */
export const server = setupServer();
