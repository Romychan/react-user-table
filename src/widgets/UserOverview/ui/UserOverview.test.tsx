import { screen, fireEvent, waitFor } from '@testing-library/react';
import { vi } from 'vitest';
import { NIL as NIL_UUID } from 'uuid';

import { UserContext, userHandlers } from '~/entities/User';

import { server } from '~/shared/api/msw/server';
import { renderWithAppProviders } from '~/shared/lib/tests';

import { UserOverview } from './UserOverview';

const mockUserStore = {
  currentUserId: NIL_UUID,
  setCurrentUserId: vi.fn(),
};

describe('UserOverview', () => {
  beforeEach(() => {
    server.use(...userHandlers);
  });

  it('should render correctly', async () => {
    renderWithAppProviders(
      <UserContext.Provider value={mockUserStore}>
        <UserOverview />
      </UserContext.Provider>,
    );

    const loaderElement = screen.getByTestId('loader');
    expect(loaderElement).toBeInTheDocument();

    const userFormElement = await screen.findByTestId('update-user-form');

    expect(loaderElement).not.toBeInTheDocument();
    expect(userFormElement).toBeInTheDocument();
  });

  it('should update user correctly', async () => {
    renderWithAppProviders(
      <UserContext.Provider value={mockUserStore}>
        <UserOverview />
      </UserContext.Provider>,
    );

    fireEvent.change(screen.getByRole('textbox', { name: 'Name' }), {
      target: { value: 'New Name' },
    });

    fireEvent.submit(screen.getByTestId('update-user'));

    await waitFor(() => expect(screen.queryAllByText('alert')).toHaveLength(0));
  });

  it('should delete user correctly', async () => {
    const spy = vi.spyOn(mockUserStore, 'setCurrentUserId');

    renderWithAppProviders(
      <UserContext.Provider value={mockUserStore}>
        <UserOverview />
      </UserContext.Provider>,
    );

    const drawerBackdropElement = screen.getByTestId('drawer-backdrop');
    const deleteUserElement = screen.getByTestId('delete-user');

    fireEvent.click(deleteUserElement);
    fireEvent.animationEnd(drawerBackdropElement);

    await waitFor(() => expect(spy).toHaveBeenCalled());
  });

  it('should the id be reset when the drawer is closed', () => {
    const spy = vi.spyOn(mockUserStore, 'setCurrentUserId');

    renderWithAppProviders(
      <UserContext.Provider value={mockUserStore}>
        <UserOverview />
      </UserContext.Provider>,
    );

    const drawerBackdropElement = screen.getByTestId('drawer-backdrop');
    const drawerCloseElement = screen.getByTestId('drawer-close');

    fireEvent.click(drawerCloseElement);
    fireEvent.animationEnd(drawerBackdropElement);

    expect(spy).toHaveBeenCalled();
  });

  it('should not be displayed if the user id is not specified', () => {
    const mockEmptyUserStore = {
      currentUserId: null,
      setCurrentUserId: vi.fn(),
    };

    renderWithAppProviders(
      <UserContext.Provider value={mockEmptyUserStore}>
        <UserOverview />
      </UserContext.Provider>,
    );

    const drawerElement = screen.queryByTestId('drawer');
    expect(drawerElement).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const container = renderWithAppProviders(
      <UserContext.Provider value={mockUserStore}>
        <UserOverview />
      </UserContext.Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
