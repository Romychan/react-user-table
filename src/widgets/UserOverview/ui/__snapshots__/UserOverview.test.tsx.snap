// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserOverview > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_backdrop_626b11 _right_626b11 _filled_626b11"
        data-testid="drawer-backdrop"
      />
      <div
        class="_container_626b11 _right_626b11 _container_d9432a"
        data-testid="drawer"
      >
        <div
          class="_header_626b11"
        >
          <div
            class="_information_626b11"
          >
            <h3
              class="_title_626b11"
            >
              Overview
            </h3>
            <p
              class="_description_626b11"
            />
          </div>
          <button
            class="_button_0302f3 _tertiary_0302f3 _xs_0302f3 _icon_0302f3"
            data-testid="drawer-close"
          >
            <svg
              class="icon icon-close"
              data-testid="icon"
              height="16"
              width="16"
            >
              <use
                xlink:href="/images/icons.svg#close"
              />
            </svg>
          </button>
        </div>
        <div
          class="_body_626b11"
        >
          <div
            class="_content_626b11"
          >
            <form
              class="_container_45f202"
              data-testid="update-user-form"
            >
              <div
                class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
              >
                <label
                  class="_label_77a9ed"
                  for="name"
                >
                  Name
                </label>
                <div
                  class="_content_77a9ed"
                >
                  <input
                    autocomplete="off"
                    class="_input_95237c _unfilled_95237c"
                    data-testid="input-field"
                    id="name"
                    name="name"
                    value="Test Name"
                  />
                </div>
              </div>
              <div
                class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
              >
                <label
                  class="_label_77a9ed"
                  for="username"
                >
                  Username
                </label>
                <div
                  class="_content_77a9ed"
                >
                  <input
                    autocomplete="off"
                    class="_input_95237c _unfilled_95237c"
                    data-testid="input-field"
                    id="username"
                    name="username"
                    value="testname"
                  />
                </div>
              </div>
              <div
                class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
              >
                <label
                  class="_label_77a9ed"
                  for="email"
                >
                  Email
                </label>
                <div
                  class="_content_77a9ed"
                >
                  <input
                    autocomplete="off"
                    class="_input_95237c _unfilled_95237c"
                    data-testid="input-field"
                    id="email"
                    name="email"
                    value="test@test.test"
                  />
                </div>
              </div>
              <div
                class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
              >
                <label
                  class="_label_77a9ed"
                  for="phone"
                >
                  Phone
                </label>
                <div
                  class="_content_77a9ed"
                >
                  <input
                    autocomplete="off"
                    class="_input_95237c _unfilled_95237c"
                    data-testid="input-field"
                    id="phone"
                    name="phone"
                    type="tel"
                    value="+7 (123) 456-78-90"
                  />
                </div>
              </div>
              <div
                class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
              >
                <label
                  class="_label_77a9ed"
                  for="city"
                >
                  City
                </label>
                <div
                  class="_content_77a9ed"
                >
                  <input
                    autocomplete="off"
                    class="_input_95237c _unfilled_95237c"
                    data-testid="input-field"
                    id="city"
                    name="city"
                    value="Moscow"
                  />
                </div>
              </div>
              <div
                class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
              >
                <label
                  class="_label_77a9ed"
                  for="address"
                >
                  Address
                </label>
                <div
                  class="_content_77a9ed"
                >
                  <input
                    autocomplete="off"
                    class="_input_95237c _unfilled_95237c"
                    data-testid="input-field"
                    id="address"
                    name="address"
                    value="Test Address"
                  />
                </div>
              </div>
              <div
                class="_actions_45f202"
              >
                <button
                  class="_button_0302f3 _danger_0302f3 _sm_0302f3"
                  data-testid="delete-user"
                  type="button"
                >
                  Delete
                </button>
                <button
                  class="_button_0302f3 _primary_0302f3 _sm_0302f3"
                  data-testid="update-user"
                  type="submit"
                >
                  Update
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      class="_container_f24d02 _bottom-center_f24d02"
      data-testid="toast-container"
    />
  </body>,
  "container": <div>
    <div
      class="_backdrop_626b11 _right_626b11 _filled_626b11"
      data-testid="drawer-backdrop"
    />
    <div
      class="_container_626b11 _right_626b11 _container_d9432a"
      data-testid="drawer"
    >
      <div
        class="_header_626b11"
      >
        <div
          class="_information_626b11"
        >
          <h3
            class="_title_626b11"
          >
            Overview
          </h3>
          <p
            class="_description_626b11"
          />
        </div>
        <button
          class="_button_0302f3 _tertiary_0302f3 _xs_0302f3 _icon_0302f3"
          data-testid="drawer-close"
        >
          <svg
            class="icon icon-close"
            data-testid="icon"
            height="16"
            width="16"
          >
            <use
              xlink:href="/images/icons.svg#close"
            />
          </svg>
        </button>
      </div>
      <div
        class="_body_626b11"
      >
        <div
          class="_content_626b11"
        >
          <form
            class="_container_45f202"
            data-testid="update-user-form"
          >
            <div
              class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
            >
              <label
                class="_label_77a9ed"
                for="name"
              >
                Name
              </label>
              <div
                class="_content_77a9ed"
              >
                <input
                  autocomplete="off"
                  class="_input_95237c _unfilled_95237c"
                  data-testid="input-field"
                  id="name"
                  name="name"
                  value="Test Name"
                />
              </div>
            </div>
            <div
              class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
            >
              <label
                class="_label_77a9ed"
                for="username"
              >
                Username
              </label>
              <div
                class="_content_77a9ed"
              >
                <input
                  autocomplete="off"
                  class="_input_95237c _unfilled_95237c"
                  data-testid="input-field"
                  id="username"
                  name="username"
                  value="testname"
                />
              </div>
            </div>
            <div
              class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
            >
              <label
                class="_label_77a9ed"
                for="email"
              >
                Email
              </label>
              <div
                class="_content_77a9ed"
              >
                <input
                  autocomplete="off"
                  class="_input_95237c _unfilled_95237c"
                  data-testid="input-field"
                  id="email"
                  name="email"
                  value="test@test.test"
                />
              </div>
            </div>
            <div
              class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
            >
              <label
                class="_label_77a9ed"
                for="phone"
              >
                Phone
              </label>
              <div
                class="_content_77a9ed"
              >
                <input
                  autocomplete="off"
                  class="_input_95237c _unfilled_95237c"
                  data-testid="input-field"
                  id="phone"
                  name="phone"
                  type="tel"
                  value="+7 (123) 456-78-90"
                />
              </div>
            </div>
            <div
              class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
            >
              <label
                class="_label_77a9ed"
                for="city"
              >
                City
              </label>
              <div
                class="_content_77a9ed"
              >
                <input
                  autocomplete="off"
                  class="_input_95237c _unfilled_95237c"
                  data-testid="input-field"
                  id="city"
                  name="city"
                  value="Moscow"
                />
              </div>
            </div>
            <div
              class="_container_77a9ed _secondary_77a9ed _left_77a9ed"
            >
              <label
                class="_label_77a9ed"
                for="address"
              >
                Address
              </label>
              <div
                class="_content_77a9ed"
              >
                <input
                  autocomplete="off"
                  class="_input_95237c _unfilled_95237c"
                  data-testid="input-field"
                  id="address"
                  name="address"
                  value="Test Address"
                />
              </div>
            </div>
            <div
              class="_actions_45f202"
            >
              <button
                class="_button_0302f3 _danger_0302f3 _sm_0302f3"
                data-testid="delete-user"
                type="button"
              >
                Delete
              </button>
              <button
                class="_button_0302f3 _primary_0302f3 _sm_0302f3"
                data-testid="update-user"
                type="submit"
              >
                Update
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
